### PR TITLE
multi table testbed

### PR DESCRIPTION
**Pull request by jerch**
Testbed and playground PR for multi table tests.

@mobiware Plz ignore the failing tests (happens due to the debug logs not matching the expected query count anymore).

### CODE REVIEW

**Comment by jerch** [comment]
@mobiware The last commit enables correct "up-pulling" of derived field values. For some reason it needs a different name resolving in `resolve_dependencies`, not sure yet why and exactly when.

**Comment by mobiware** [comment]
@jerch yes I was about to say that the initial commit didn’t address the parent model defining computed fields depending on the children models, but your latest commit seems to take care of this.

Looks like I took the wrong approach on my own PR, i.e. treat parent and child models as 2 different models with their own disjointed sets of computed fields.

I’ll run my own test cases against this branch a bit later today but this sure looks promising, thanks !

**Comment by jerch** [comment]
@mobiware Found the reason for the weird attr access for inherited o2o relations - it is basically the same reason, why you added the `field.model == model` check in the generator.

Imho this needs simply a correction of paths in depends rules - while the path is valid on the model, where the field was declared on (base model as in `field.model`), the derived one leads to invalid path lookups. But we can patch that in `resolve_dependencies` to the correct lookup before descending into deeper field resolving. This has to be done for any derived model in this way:

```
# model inheritance chain
Base --> DerivedA --> DerivedB --> DerivedC

# depends on cf on base
['deriveda.derivedb.derivedc', [...]]


# turns into these deps for cf on model
Base: deriveda.derivedb.derivedc
DerivedA: derivedb.derivedc
DerivedB: derivedc
DerivedC: self
```
Programmatically we can start at `field.model` and descent the relation path until we found the model at hand. The remaining relation path is the correct one for the derived model at hand and can be used for further field resolving.

If the current model is not part of the relation path of the depends rule (might happen from deps to other derived models), we can simply skip that rule.

**Comment by mobiware** [comment]
Finally got a chance to run my tests on your branch, and added a few tests too:
```
# Test classes for multi table inheritance support
class ParentModel(ComputedFieldsModel):
    x = models.IntegerField(default=0)
    y = models.IntegerField(default=0)

    @computed(models.IntegerField(default=0), depends=[["self", ["x", "y"]]])
    def z(self):
        return self.x + self.y

    @computed(models.CharField(max_length=255, null=True, blank=True), depends=[["childmodel", ["username"]], ["childmodel2", ["pseudo"]]])
    def name(self):
        if hasattr(self, "childmodel"):
            return self.childmodel.username
        elif hasattr(self, "childmodel2"):
            return self.childmodel2.pseudo


class ChildModel(ParentModel):
    username = models.CharField(max_length=255, default="")

    a = models.IntegerField(default=0)
    b = models.IntegerField(default=0)

    @computed(models.IntegerField(default=0), depends=[["self", ["a", "b"]]])
    def c(self):
        return self.a + self.b


class ChildModel2(ParentModel):
    pseudo = models.CharField(max_length=255, default="")

    @computed(models.CharField(max_length=255, null=True, blank=True), depends=[["parentmodel_ptr", ["name", "z", "x"]]])
    def other_name(self):
        return f"{self.x}{self.name}{self.z}"


class DependsOnParent(ComputedFieldsModel):
    parent = models.ForeignKey(ParentModel, on_delete=models.CASCADE)

    @computed(models.IntegerField(default=0), depends=[["parent", ["x"]]])
    def x2(self):
        return self.parent.x * 2


class DependsOnParentComputed(ComputedFieldsModel):
    parent = models.ForeignKey(ParentModel, on_delete=models.CASCADE)

    @computed(models.IntegerField(default=0), depends=[["parent", ["z"]]])
    def z2(self):
        return self.parent.z * 2
```

```
class MultiTableInheritanceModel(TestCase):
    def test_depends_on_child_model(self):
        child1 = ChildModel.objects.create(username="Child")
        child1.refresh_from_db()
        self.assertEqual(child1.name, "Child")

        child1.username = "Kid"
        child1.save()

        child1.refresh_from_db()

        self.assertEqual(child1.name, "Kid")

        child2 = ChildModel2.objects.create(pseudo="Pseudo")
        child2.refresh_from_db()
        self.assertEqual(child2.name, "Pseudo")

        child2.pseudo = "Kid"
        child2.save()

        child2.refresh_from_db()

        self.assertEqual(child2.name, "Kid")

    def test_parent_depends_on_self(self):
        child = ChildModel.objects.create(username="Child", x=12, y=67)
        child.refresh_from_db()
        self.assertEqual(child.name, "Child")
        self.assertEqual(child.z, 12+67)

        child.y = 8
        child.save()

        child.refresh_from_db()

        self.assertEqual(child.z, 20)

    def test_child_depends_on_self(self):
        child = ChildModel.objects.create(username="Child", a=12, b=67)
        child.refresh_from_db()
        self.assertEqual(child.name, "Child")
        self.assertEqual(child.c, 12+67)

        child.b = 8
        child.save()

        child.refresh_from_db()

        self.assertEqual(child.c, 20)

    def test_child_depends_on_self_and_parent(self):
        child = ChildModel2.objects.create(pseudo="Child", x=12, y=67)
        child.refresh_from_db()
        self.assertEqual(child.name, "Child")
        self.assertEqual(child.z, 12+67)
        self.assertEqual(child.other_name, "12Child79")

        child.x = 10
        child.save()

        child.refresh_from_db()

        self.assertEqual(child.z, 77)
        self.assertEqual(child.other_name, "10Child77")

    def test_other_class_depends_on_parent(self):
        parent = ParentModel.objects.create(x=3, y=5)
        self.assertEqual(parent.z, 8)
        self.assertIsNone(parent.name)

        other = DependsOnParent.objects.create(parent=parent)
        self.assertEqual(other.x2, 6)

        parent.x = 7
        parent.save()

        other.refresh_from_db()

        self.assertEqual(parent.z, 12)
        self.assertEqual(other.x2, 14)

    def test_other_class_depends_on_parent_computed(self):
        parent = ParentModel.objects.create(x=3, y=5)
        self.assertEqual(parent.z, 8)
        self.assertIsNone(parent.name)

        other = DependsOnParentComputed.objects.create(parent=parent)
        self.assertEqual(other.z2, 16)

        parent.x = 7
        parent.save()

        other.refresh_from_db()

        self.assertEqual(parent.z, 12)
        self.assertEqual(other.z2, 24)

    def test_other_class_depends_on_parent_child(self):
        child = ChildModel.objects.create(username="Child", x=3, y=5)
        child.refresh_from_db()
        self.assertEqual(child.z, 8)
        self.assertEqual(child.name, "Child")

        other = DependsOnParent.objects.create(parent=child)
        self.assertEqual(other.x2, 6)

        child.x = 7
        child.save()

        other.refresh_from_db()

        self.assertEqual(child.z, 12)
        self.assertEqual(other.x2, 14)

    def test_other_class_depends_on_parent_computed_child(self):
        child = ChildModel.objects.create(username="Child", x=3, y=5)
        child.refresh_from_db()
        self.assertEqual(child.z, 8)
        self.assertEqual(child.name, "Child")

        other = DependsOnParentComputed.objects.create(parent=child)
        self.assertEqual(other.z2, 16)

        child.x = 7
        child.save()

        other.refresh_from_db()

        self.assertEqual(child.z, 12)
        self.assertEqual(other.z2, 24)
```

These are the tests from my branch, and I added 4 tests to test other models having dependencies with the Parent model in a multi-table inheritance hierarchy.

When the parent model object has no child, it works fine (see `test_other_class_depends_on_parent` and `test_other_class_depends_on_parent_computed`) but when that parent model has a child, I get an error (the last assert of both `test_other_class_depends_on_parent_child` and `test_other_class_depends_on_parent_computed_child` fails, the computed field on `other` hasn't been recomputed)

**Comment by jerch** [comment]
@mobiware Yeah, the current hotfix is not the right one, it just works "by accident". We basically have to apply the correct relation path translations as described above. Will need abit for that and proper tests (might also copy over your tests for this), as its cumbersome to cover all edge cases.

**Comment by mobiware** [comment]
@jerch sure, please copy those tests over. 6 of them are already passing, only the last 2 are failing

**Comment by jerch** [comment]
@mobiware Halfway done with the rewrite. It now pulls the right depends strings for the model inheritance descent, even over multiple classes.

Your last 2 test cases are still failing, as they need proper ascending of updated fields to their source model (they refer to models higher up in the inheritance chain), which is currently not yet reflected by the graph (it only forwards field updates on the model at hand currently). Still need to find a good way to propagate that back to the higher model level.

File: computedfields/graph.py [hunk 1]
```diff
@@ -437,20 +437,42 @@ def resolve_dependencies(self, computed_models):
...
-                    print('special case multi table inheritance', model, field)
-
                 fieldentry = global_deps.setdefault(model, {}).setdefault(field, {})
                 local_deps.setdefault(model, {}).setdefault(field, set())
 
                 depends = real_field._computed['depends']
 
+                # fields contributed from multi table model inheritance need patched depends rules,
+                # so the relation paths match the changed model entrypoint
+                if real_field.model != model and not real_field.model._meta.abstract:
```
Comment (mobiware) [inline computedfields/graph.py:447]: just out of curiosity, why do you have to add the `and not real_field.model._meta.abstract` condition ? From my observations it seemed like fields inherited from an abstract model have their `model` attribute pointing to the Concrete model, not the Abstract model.
So Abstract model fields shouldn't even pass `if real_field.model != model`
In fact, IIRC the abstract model doesn't even trigger its own `class_prepared` signal

Comment (jerch) [reply computedfields/graph.py:447]: Yes, it is just a security measure to somewhat cover an implicit behavior here more explicit:
- fields from decoration stage still contain the real source model (thus an abstract model)
- in `models_with_computedfields` we pull their correct model association by equality check, this works correctly, which is abit surprising, as the same (=equal) field pulled from `model._meta.fields` vs. from decoration stage has a different `.model` attribute

I have not looked up django's code regarding filling in the `field.model` attribute, so this here is just a guard to not grab here a totally unkown abstract model by accident.

> In fact, IIRC the abstract model doesn't even trigger its own class_prepared signal

Yeah the resolver doesn't know anything about abstract models, as we only listen for models from that signal. Thats quite different to the old metaclass approach, here those qorkaround for stracts were needed.

Comment (mobiware) [reply computedfields/graph.py:447]: OK thanks for the explanation.

> which is abit surprising, as the same (=equal) field pulled from model._meta.fields vs. from decoration stage has a different .model attribute

Makes sense when looking at Django's `Field` `__eq__` operator implementation:
```
    def __eq__(self, other):
        # Needed for @total_ordering
        if isinstance(other, Field):
            return self.creation_counter == other.creation_counter
        return NotImplemented
```

**Comment by jerch** [comment]
@mobiware The last commit contain a quick hack to see whether with field ascending we can cover deps to higher models from the inheritance chain. This also passes your last 2 tests. So the idea works, but again will take some time to coded and integrated correctly. Also we will have to write more tests again, as the inheritance is like another dimension on top of the resolving needs.

**Comment by jerch** [comment]
@mobiware  Oh well, this gets indeed abit exhaustive, and not sure yet if we can get it race condition free. Trying to illustrate it:

```
# inherited models
A(Model)    -->    B(A)    -->    C(B)

# some fields
.a                .b              .c

# fields on instance/qs (can be depended on)
.a                .a              .a
                  .b              .b
                                  .c
```
This leads to the fact, that an incoming update changeset in `update_dependent` as `Model{changed_fields}` does not map directly to a list of `Dependent{cfs_to_update}` anymore, but has to be calculated from the inheritance chain.

For `update_fields=None` (default case) this gets interesting:
- update on `A{None}` --> `A{None}` + `B{fields_from_A}` + `C{fields_from_A}`
- update on `B{None}` --> `A{None}` + `B{None}` + `C{fields_from_B}`
- update on `C{None}` --> `A{None}` + `B{None}` + `C{None}`

For a single field it depends, where the field was defined:
- update `A{a}` --> `A{a}` + `B{a}` + `C{a}`
- update `B{b}` --> `B{b}` + `C{b}` (no A rule)
- update `C{c}` --> `C{c}` (no A rule, no B rule)
- update `C{b}` --> `B{b}` + `C{b}` (no A rule)

Putting this all together, mixed fields in `update_fields` can be resolved like that:
- update `B{a,b}` --> `A{a}` + `B{a,b}` + `C{a,b}`

This will perform really bad if done at runtime in `_queryset_for_update` over and over, thus we have to find a way to move it into the bootstrapping part. Ideal solution would be to already populate it into the graph (as single source of deps truth), will see if thats feasible as it involves copying subtrees alot.

**Comment by mobiware** [comment]
@jerch looks complicated indeed. What about my initial approach, i.e. treat `A(Model)` and `B(A)` as 2 completely different models with a O2O relationship joining them ? We'd only have to change dependencies a bit when building the graph i.e. if `B.z` depends on `self.x` then we change that dependency to `a_ptr.x` when building the graph

**Comment by jerch** [comment]
> looks complicated indeed. What about my initial approach, i.e. treat A(Model) and B(A) as 2 completely different models with a O2O relationship joining them ? We'd only have to change dependencies a bit when building the graph i.e. if B.z depends on self.x then we change that dependency to a_ptr.x when building the graph

I am abit afraid of the additional costs at runtime for local field discovery, that live on a parent table. We would have to handle any field in depends rules strictly on their source model with `update_dependent`, while the ORM already populates those fields by its joints on derived models. This results in tons of overfetching. Or speaking in graph terms - it deepens the graph, while handling them locally flattens it. And the latter is much easier to optimize:
Imagine the 3 models above plus one computed field `C.comp`, that depends on all 3 fields as in `depends=[['self', ['a', 'b', 'c']]]`.
The flat model access could handle the update of `C{None}` by one select query, and one update query, while fields are recalculated locally. Thats pulling one matrix of data in _[Cfields x records]_ and updating one row in that matrix. Or for the single instance - updates `C.comp` right away before saving the instance.

For the nested model access it gets more cumbersome:
A single instance would be saved with updating `C.comp` from `C.c` only (now dirty), additionally `update_dependent` called from `post_save` signal would see `a_ptr.a, b_ptr.b` as changed, thus pulling C records in `a_ptr, b_ptr` again (the same records we had before, duh) to do the final update of `C.comp`. Btw - the "C records in `a_ptr, b_ptr`" is a relatively expensive query, as it is a `C.objects.filter(a_ptr__in=[As_in_initial_qs], b_ptr=[Bs_in_initial_qs])`.

Imho the needed upcast / downcast between A/B/C objects is hard to cover correctly with either way - there are always edge cases we have to handle explicitly on graph level (as I already started with the path shortening). Here I wonder whether we could simply pull the field deps across, thus the graph already knows that an update of `A.a` / `B.a` / `C.a` triggers recalculation of `C.comp` (as pulled from inheritance `C.a` --> `B.a` --> `A.a`). It would bloat the graph and the maps, but mostly make it broader, not much deeper. Which again is easier to optimize.

**Comment by jerch** [comment]
@mobiware I think we should place a note in the docs about multi table inheritance not being supported until we have resolved this.

**Comment by jerch** [comment]
@mobiware Btw - most of the things regarding multi table inheritance should already work, if someone places the right `*_ptr` paths in _depends_. If correctly done, it is merely an o2o relation (again with the ugly `update_dependent` loops).
Example:
```python

class Base(models.Model):
  basename = models.CharField(max_length=32)

class Derived(Base):
  @computed(models.CharField(max_length=32), depends=[
    ['self', ['basename']],     # catches updates from Derived{basename}
    ['base_ptr', ['basename']]  # catches updates from Base{basename}
  ])
  def comp(self):
    return self.basename

# Derived.comp correctly updates from either
derived_instance.save()  # or with update_fields=['basename']
# or
base_instance.save()   # or again with update_fields=['basename']
# or even from mass actions:
Base.objects.all().update(basename='new value')
update_dependent(Base.objects.all(), update_fields=['basename'])
```
Ofc those `*_ptr` paths are hard to set up correctly, very likely to miss relations here.

**Comment by jerch** [comment]
@mobiware Indeed, with proper additional depends rules also your last 2 tests are working (see last commit).

About automatically extending the graph by those rules:
- updates from higher model types need additional `*_ptr` rules (ascending catch rule)
- updates from lower models need the reverse path of `*_ptr` (descending catch rule)

**Comment by mobiware** [comment]
@jerch sounds good, even if the `depends` rules like `["parent.childmodel", ["x"]]` are counter-intuitive, at least it makes it possible to use DCF on Multi-Table models, thanks !

**Comment by jerch** [comment]
@mobiware What do you think - should we try to place those additional rules automatically, or just state clearly in the docs, how to do it manually?

**Comment by mobiware** [comment]
@jerch I guess we can add those in the doc as a first step, and hopefully find a way in a subsequent PR to generate them automatically from more "natural" rules like `["self", ["x"]]`

**Comment by jerch** [comment]
@mobiware Added a section in examples about multi table support and how to derive the counter-intuitive depends rules. Plz check the construction rules, the wording and whether they can be followed easy.

File: docs/examples.rst [hunk 2]
```diff
@@ -309,6 +309,174 @@ limited for m2m fields. Also see below under optimization examples.
...
+-----------------------
+
+.. |br| raw:: html
+
+   <br />
+
+
+Multi table inheritance works with computed fields with some restrictions you have to be aware of.
+The following requires basic knowledge about multi table inheritance in Django and its similarities
+to o2o relations on accessor level (also see official Django docs).
```
Comment (mobiware) [inline docs/examples.rst:322]: nit: maybe a link to the Django doc ?
https://docs.djangoproject.com/en/3.0/topics/db/models/#multi-table-inheritance

**Comment by jerch** [comment]
@mobiware Oh dang my "up-pulling" example would not work - name clash :smile_cat:.

**Comment by jerch** [comment]
@mobiware Added the link, and changed the up-pulling to something else. Ready to merge?

**Comment by mobiware** [comment]
@jerch sorry, missed that error in the doc. Yes, LGTM
